silence/silence.go

```diff
@@ -276,6 +325,11 @@ func (s *Silences) setSilence(sil *pb.Silence) error {
 	if err != nil {
 		return err
 	}
+
+	_, err = s.mc.Add(sil)
+	if err != nil {
+		return err
+	}
```

Comment:
This is not sufficient. This code path is executed when a silence is set in this Alertmanager instance.
When we receive one set in a peer Alertmanager, it is directly inserted into the state.

To work with the mesh protocol we call `gossipData.Merge()` with the update we construct below. Inserting into our own state (`gossipData`) also happens in that method. When we receive updates from other peers `gossipData.mergeDelta()` is called (via `GossipBroadcast()` and `OnGossip()`) which also changes our state. For the latter this code here is not executed and hence we end up with a silence that's not in our cache.
This would require access to the cache in `gossipData.mergeDelta()`, which we currently don't have. Maybe it would be wise to make `type gossiper` implement the `mesh.Gossiper` and `mesh.GossipData` interface at once?

Secondly, we store snapshots of silence data that we can restore from after a binary restart. In `loadSnapshot()` we have to populate the cache again for the silences we load.